cmd/registry/cmd/delete/delete.go

```diff
@@ -30,145 +31,332 @@ import (
 		},
 	}
 
 	cmd.Flags().StringVar(&filter, "filter", "", "Filter selected resources")
 	cmd.Flags().IntVar(&jobs, "jobs", 10, "Number of actions to perform concurrently")
+	cmd.Flags().BoolVarP(&force, "force", "f", false, "Force deletion of child resources")
 	return cmd
 }
 
-type deleteTask struct {
-	client       connection.RegistryClient
-	resourceName string
-	resourceKind string
+type deleteHandler struct {
+	ctx         context.Context
+	client      connection.RegistryClient
+	adminClient connection.AdminClient
+	name        string
+	filter      string
+	force       bool
+	count       int
+	taskQueue   chan<- core.Task
 }
 
-func (task *deleteTask) String() string {
-	return "delete " + task.resourceName
-}
+func (h *deleteHandler) traverse() error {
+	// Define aliases to simplify the subsequent code.
+	name := h.name
+	ctx := h.ctx
+	client := h.client
+	adminClient := h.adminClient
+	filter := h.filter
 
-func (task *deleteTask) Run(ctx context.Context) error {
-	log.Debugf(ctx, "Deleting %s %s", task.resourceKind, task.resourceName)
-	switch task.resourceKind {
-	case "api":
-		return task.client.DeleteApi(ctx, &rpc.DeleteApiRequest{Name: task.resourceName})
-	case "version":
-		return task.client.DeleteApiVersion(ctx, &rpc.DeleteApiVersionRequest{Name: task.resourceName})
-	case "spec":
-		return task.client.DeleteApiSpec(ctx, &rpc.DeleteApiSpecRequest{Name: task.resourceName})
-	case "artifact":
-		return task.client.DeleteArtifact(ctx, &rpc.DeleteArtifactRequest{Name: task.resourceName})
-	default:
-		return nil
+	// First try to match collection names.
+	if project, err := names.ParseProjectCollection(name); err == nil {
+		return core.ListProjects(ctx, adminClient, project, filter, h.projectHandler())
+	} else if api, err := names.ParseApiCollection(name); err == nil {
+		return core.ListAPIs(ctx, client, api, filter, h.apiHandler())
+	} else if deployment, err := names.ParseDeploymentCollection(name); err == nil {
+		return core.ListDeployments(ctx, client, deployment, filter, h.apiDeploymentHandler())
+	} else if rev, err := names.ParseDeploymentRevisionCollection(name); err == nil {
+		return core.ListDeploymentRevisions(ctx, client, rev, filter, h.apiDeploymentRevisionHandler())
+	} else if version, err := names.ParseVersionCollection(name); err == nil {
+		return core.ListVersions(ctx, client, version, filter, h.apiVersionHandler())
+	} else if spec, err := names.ParseSpecCollection(name); err == nil {
+		return core.ListSpecs(ctx, client, spec, filter, false, h.apiSpecHandler())
+	} else if rev, err := names.ParseSpecRevisionCollection(name); err == nil {
+		return core.ListSpecRevisions(ctx, client, rev, filter, false, h.apiSpecRevisionHandler())
+	} else if artifact, err := names.ParseArtifactCollection(name); err == nil {
+		return core.ListArtifacts(ctx, client, artifact, filter, false, h.artifactHandler())
+	}
+
+	// Then try to match resource names containing wildcards, these also are treated as collections.
```

Comment:
Hmm. Should this be part of the collections' regexp?

Reply:
I think we might benefit from keeping these cases separate -- currently when the collections regexp matches, we know that the last segment is the collection name. That wouldn't be true if we merged these cases, and I'm not sure the more powerful regexp would be worth its complexity.